mantidproject/mantid
# Eng Diff UI load and average logs when loading runs

**Description of work.**

These changes allow the user to select the logs of interest in the settings and when focussed runs are loaded in the fitting tab of the UI, a grouped workspace is created that contains a table per log (each run is a row) that holds the averaged and standard deviation of the proton charge weighted log value of that run. There is an additional table in the grouped workspace that summarises the run numbers and run info (current, title etc.). The grouped workspace structure chosen is based on the existing hdf5 file structure exported in the legacy Open Genie software. The user selected logs are saved to the mantid settings and will be set next time the interface is opened. 

The rows in the log tables are ordered in the same way as the rows in the table inside the fitting tab - removing a tracked workspace in the fitting tab removes a row from each of the log tables. To do this there are changes to the behaviour when a user attempts to load a run that has already been loaded:  if it is tracked (i.e. is already a corresponding row in the fitting tab table) then no action will be taken, if it is untracked it will be retrieved from the ADS (the UI will remember any background subtraction etc.) and a row is added to the various tables. Previously the user was allowed to load in any number of duplicate runs - this was possibly an oversight.



**To test:**

(1) Follow the instructions in PR#29130 to load focused data in the fitting tab of the Eng Diff UI
(2) Loading the runs should also produce a group of table workspaces in the ADS (containing a table that summarises the run info and then one table per log). Each table has a row per run loaded.
(3) Check that removing a run in the Fitting tab table also removes the correct row from the log tables.
(4) Removing all (either by removing the last run or clicking 'remove all') should delete the log tables
(5) Check you cannot add a focused run with the same xunit more than once (change of behaviour to make updating tables easier)
(6) Check that if you remove a run in the fitting tab and then load it again it doesn't load the data but rather retrieves it from the existing workspace in the ADS (test this by removing the background and playing around with the background subtraction params - it should remember them)
(7) Remove all workspaces and go to the settings (cog button in  bottom left) and change the logs to load. Load a run and check the new logs are loaded.
(8) With one run loaded, go to the settings again and change the logs to load. If you load another run the old selected logs will still be loaded. If you close mantid and re-open it it should remember the logs you selected in the settings and will load these when you next load files

Relates to (but does not resolve/fix) #28942 (all the meta-data/logs is to be exported/saved with the final fits)

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
